### PR TITLE
Gate scholarship form page

### LINKED ISSUE #245: Gate scholarship form page behind authenticated users only
**Is your feature request related to a problem? Please describe.**
Currently anyone on the internet can access our scholarship form at https://dreamerscholars.web.app/scholarships/new. 

**Describe the solution you'd like**
We should gate the form page so that only authenticated users can access/see it. We'll want to either redirect users or bring up the login dialog first. Alternatively we can show some sort of message that they need to log in first. We could also make the **Add** link in the header disabled with mouse hover text saying you need to be logged in to add scholarships.

**Additional context**
React Router has an [example auth workflow](https://reactrouter.com/web/example/auth-workflow) on how to redirect going to protected pages.

#178 is a related issue as Firestore security rules can deny write access to unauthenticated users.

### PR DESCRIPTION
**Pull request by edwingl10**
## Description

Gate the scholarship form page from users who are not authenticated. 

## Motivation and Context

Fixes #245 

## How Has This Been Tested?

On laptop using firefox. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] User visible change (users will notice UI or functional changes)

## Previewing Changes

1. Click the `https://s4us-pr-NUMBER.onrender.com` URL **render[bot]** left in the comment below.
2. Make sure you are signed out
3. Try to access the scholarship form page by clicking on 'Add' in the header

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: src/components/LoginButton.jsx [hunk 1]
```diff
@@ -16,26 +17,29 @@ export default function LoginButton() {
...
-        Sign out
+  return isSignedIn ? (
+    <ProfileMenu
+      signOut={signUserOut}
+      name={user.displayName}
+      email={user.email}
+      photo={user.photoURL}
+    />
+  ) : (
+    <>
```
Comment (josueetcom) [inline src/components/LoginButton.jsx:31]: These aren't needed when there's a single component btw

File: src/components/ProtectedRoute.jsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <Route
+      path
+      render={() => {
+        return isSignedIn ? (
+          <Component />
+        ) : (
+          <Redirect
+            to={{
+              pathname: `/`,
```
Comment (josueetcom) [inline src/components/ProtectedRoute.jsx:29]: This should be whatever path the user's currently on or `/` if there's no history.

File: src/components/LoginDialog.jsx [hunk 3]
```diff
@@ -30,7 +40,7 @@ export default function LoginDialog() {
 
   return (
     <Dialog
-      open={params.get('login') === 'true' || false}
+      open={(params.get('login') === 'true' && !isSignedIn) || false}
```
Comment (josueetcom) [inline src/components/LoginDialog.jsx:43]:
```suggestion
      open={params.get('login') === 'true' && !isSignedIn}
```

File: src/components/LoginDialog.jsx [hunk 4]
```diff
@@ -11,6 +11,16 @@ export default function LoginDialog() {
...
   const history = useHistory();
   const params = new URLSearchParams(location.search);
+  const [isSignedIn, setIsSignedIn] = useState(!!firebase.auth().currentUser);
+
+  useEffect(() => {
+    const unregisterAuthObserver = firebase
+      .auth()
+      .onAuthStateChanged((user) => {
+        setIsSignedIn(!!user);
+      });
```
Comment (josueetcom) [inline src/components/LoginDialog.jsx:21]:
```suggestion
      .onAuthStateChanged(user => setIsSignedIn(!!user));
```

File: src/App.jsx [hunk 5]
```diff
@@ -70,13 +79,19 @@ function App() {
...
+            <Header />
+            <Switch>
+              {routes.map(({ path, component, title, guard }) => (
+                <RouteWithTitle
+                  key={path}
+                  {...{ path, component, title, guard }}
+                />
+              ))}
+            </Switch>
+            <Route path="/" component={LoginDialog} />
```
Comment (josueetcom) [inline src/App.jsx:92]: What does this do? We already have a route for this path on line 66.

File: src/components/LoginButton.jsx [hunk 6]
```diff
@@ -16,10 +15,9 @@ export default function LoginButton() {
         setIsSignedIn(!!user);
       });
     return unregisterAuthObserver;
-  }, []); // [] skips cleanup of this effect until the component is unmounted
+  }, []);
 
-  const signUserOut = () => firebase.auth().signOut() && handleClose(); // sign out + closes dialog
-  const showLoginDialog = () => setLoginDialogOpen(true);
+  const signUserOut = () => firebase.auth().signOut(); // sign out + closes dialog
```
Comment (josueetcom) [inline src/components/LoginButton.jsx:20]:
```suggestion
  const signUserOut = () => firebase.auth().signOut();
```

**Review by josueetcom — CHANGES_REQUESTED** [review]
Some UI behavior I noticed:
- If visit the page and log in, then I have to press back 3 times to get to the previous site. This is probably too much considering that the page itself doesn't change.
- If I'm already logged in and then go to a page with `&login=true` in the URL, then the Login dialog shows up even though I'm logged in.
- If I'm elsewhere besides the home page and then click on Login then it redirects to the home page where the login dialog shows up.

These issues need to be addressed. Also since it's tricky it'd be best to write test cases for these redirects. The React Router library has ways to [start at specific routes](https://reactrouter.com/web/guides/testing/starting-at-specific-routes) and [check location in tests](https://reactrouter.com/web/guides/testing/checking-location-in-tests).

Instead of `login=true` an easier to do this might be to have a `/login` route with the login dialog body that redirects to `/` or whatever path you were trying to access. I think often times I see something like `referrer=` or `nextPage=` query parameters in the URLs.

File: src/App.jsx (hunk 5 above)
Comment (edwingl10) [reply src/App.jsx:92]: This is so that the login dialog appears, I couldn't get it to show without it making it a route.

File: src/components/ProtectedRoute.jsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        ) : (
+          <Redirect
+            to={{
+              pathname: lastLocation?.pathname || '/',
+              search: '?login=true',
+              state: { from: path },
+            }}
+          />
+        );
+      }}
```
Comment (josueetcom) [inline src/components/ProtectedRoute.jsx:32]:
```suggestion
        return isSignedIn === false ? (
          <Redirect
            to={{
              pathname: lastLocation?.pathname || '/',
              search: '?login=true',
              state: { from: path },
            }}
          />
        ) : (
          <Component />
        );
      }}
```

File: src/App.jsx [hunk 8]
```diff
@@ -14,24 +14,32 @@ import ScholarshipForm from './pages/ScholarshipForm';
...
+import ProtectedRoute from './components/ProtectedRoute';
+import LoginDialog from './components/LoginDialog';
 
-function RouteWithTitle({ path, component, title }) {
+function RouteWithTitle({ path, component, title, guard }) {
   useEffect(() => {
     document.title = `${BRAND_NAME} | ${title}`;
   }, [title]);
+  if (guard) return <ProtectedRoute {...{ path, component }} />;
   return <Route {...{ path, component }} />;
```
Comment (josueetcom) [inline src/App.jsx:25]:
```suggestion
  return guard ? <ProtectedRoute {...{ path, component }} /> : <Route {...{ path, component }} />;
```

File: src/components/LoginDialog.jsx [hunk 9]
```diff
@@ -1,32 +1,36 @@
...
-    signInSuccessWithAuthResult: () => false,
-  },
-};
+export default function LoginDialog() {
+  const location = useLocation();
+  const { showLoginDialog } = location.state || { showLoginDialog: false };
+  const history = useHistory();
+
+  const closeDialog = () =>
+    history.replace({ state: { showLoginDialog: false } });
```
Comment (josueetcom) [inline src/components/LoginDialog.jsx:16]: I suggest ordering/grouping these so that related statements are together, e.g.
```js
  const location = useLocation();
  const { showLoginDialog } = location.state || { showLoginDialog: false };
  
  const history = useHistory();
  const closeDialog = () =>
    history.replace({ state: { showLoginDialog: false } });

```

The extra blank line at the bottom is so that `uiConfig` is separated from relatively unrelated lines.

File: src/components/LoginButton.jsx (hunk 6 above)
Comment (josueetcom) [reply src/components/LoginButton.jsx:15]: Still relevant.

File: src/components/LoginButton.jsx [hunk 10]
```diff
@@ -30,11 +23,15 @@ export default function LoginButton() {
...
+    <Button
+      variant="contained"
+      color="primary"
+      component={Link}
+      to={{
+        state: { showLoginDialog: true },
+      }}
+      replace>
+      Login
+    </Button>
```
Comment (josueetcom) [inline src/components/LoginButton.jsx:35]: This is not a comment to address in this PR, but it's interesting how simple this `Button` has become. It makes me think we should define this as `LoginButton` instead and have `HeaderNavMenu` handle the question _should I show a dropdown or a Login button?_

**Review by josueetcom — CHANGES_REQUESTED** [review]
Just some minor changes

File: src/components/LoginDialog.jsx [hunk 11]
```diff
@@ -1,32 +1,36 @@
...
-    firebase.auth.FacebookAuthProvider.PROVIDER_ID,
-  ],
-  credentialHelper: 'none', // hacky way to disable redirect on email login
-  callbacks: {
-    signInSuccessWithAuthResult: () => false,
-  },
-};
+export default function LoginDialog() {
+  const location = useLocation();
+  const { showLoginDialog } = location.state || { showLoginDialog: false };
```
Comment (josueetcom) [inline src/components/LoginDialog.jsx:12]: For consistency maybe the default here should be `undefined` as well. It's treated as a falsy value too so I think the login dialog won't show up by default.

**Comment by josueetcom** [comment]
@gonzalo1330 I think it'd be helpful for you to see the changes in this PR and understand them.

Some points of interest:
- Notice how to show/hide the dialog we're now depending on a [`location`](https://reactrouter.com/web/api/location)'s `state.showLoginDialog`. Note how it can be `true`, `false` or `undefined`.
- Notice how [`history.replace`](https://reactrouter.com/web/api/history), [`Link.replace`](https://reactrouter.com/web/api/Link/replace-bool) and [`Redirect.push`](https://reactrouter.com/web/api/Redirect/push-bool) are used to _replace_ the current page (so that a new browser history entry isn't created).
- Notice how `ProtectedRoute` distinguishes between the three User states: _definitely signed in_, _definitely not signed in_, and _unknown_. It shows a progress while the state is unknown instead of treating unknown as false and doing a redirect immediately.

File: src/components/LoginButton.jsx (hunk 10 above)
Comment (edwingl10) [reply src/components/LoginButton.jsx:35]: Yeah I see how simple it is. We can make another PR where we move the login to the `HeaderNavMenu`

File: src/components/LoginDialog.jsx (hunk 11 above)
Comment (edwingl10) [reply src/components/LoginDialog.jsx:12]: yup works the same.

**Review by josueetcom — APPROVED** [review]
